File: ios-app/AnagramArena/Sources/Core/Engine/LetterGenerator.swift
import Foundation

struct WeightedLetterPool {
    private let weightedEntries: [(letter: Character, weight: Int)]
    private let totalWeight: Int

    init(weights: [Character: Int]) {
        self.weightedEntries = weights
            .filter { $0.value > 0 }
            .sorted(by: { $0.key < $1.key })
            .map { (letter: $0.key, weight: $0.value) }
        self.totalWeight = weightedEntries.reduce(0) { $0 + $1.weight }
    }

    func draw<R: RandomNumberGenerator>(using generator: inout R) -> Character {
        precondition(totalWeight > 0, "Weighted pool cannot be empty")

        let target = Int.random(in: 0..<totalWeight, using: &generator)
        var running = 0

        for entry in weightedEntries {
            running += entry.weight
            if target < running {
                return entry.letter
            }
        }

        return weightedEntries[weightedEntries.count - 1].letter
    }

    func contains(_ letter: Character) -> Bool {
        weightedEntries.contains(where: { $0.letter == letter.uppercased().first })
    }

    var letters: Set<Character> {
        Set(weightedEntries.map(\.letter))
    }
}

enum DefaultLetterWeights {
    static let vowels: [Character: Int] = [
        "A": 15,
        "E": 21,
        "I": 13,
        "O": 13,
        "U": 5
    ]

    static let consonants: [Character: Int] = [
        "B": 2,
        "C": 3,
        "D": 6,
        "F": 2,
        "G": 3,
        "H": 2,
        "J": 1,
        "K": 1,
        "L": 5,
        "M": 4,
        "N": 8,
        "P": 4,
        "Q": 1,
        "R": 9,
        "S": 9,
        "T": 9,
        "V": 1,
        "W": 1,
        "X": 1,
        "Y": 1,
        "Z": 1
    ]
}

struct LetterGenerator {
    private let vowelPool: WeightedLetterPool
    private let consonantPool: WeightedLetterPool

    init(
        vowelWeights: [Character: Int] = DefaultLetterWeights.vowels,
        consonantWeights: [Character: Int] = DefaultLetterWeights.consonants
    ) {
        self.vowelPool = WeightedLetterPool(weights: vowelWeights)
        self.consonantPool = WeightedLetterPool(weights: consonantWeights)
    }

    func generateLetter<R: RandomNumberGenerator>(kind: LetterKind, using generator: inout R) -> Character {
        switch kind {
        case .vowel:
            return vowelPool.draw(using: &generator)
        case .consonant:
            return consonantPool.draw(using: &generator)
        }
    }

    func isValidLetter(_ letter: Character, for kind: LetterKind) -> Bool {
        switch kind {
        case .vowel:
            return vowelPool.contains(letter)
        case .consonant:
            return consonantPool.contains(letter)
        }
    }
}

File: android-app/core/src/main/kotlin/com/bhan796/anagram/core/engine/LetterGenerator.kt
package com.bhan796.anagram.core.engine

import com.bhan796.anagram.core.model.LetterKind
import kotlin.random.Random

class WeightedLetterPool(weights: Map<Char, Int>) {
    private val weightedEntries: List<Pair<Char, Int>> =
        weights
            .filterValues { it > 0 }
            .toList()
            .sortedBy { it.first }

    private val totalWeight: Int = weightedEntries.sumOf { it.second }

    fun draw(random: Random): Char {
        require(totalWeight > 0) { "Weighted pool cannot be empty" }

        val target = random.nextInt(totalWeight)
        var running = 0

        for ((letter, weight) in weightedEntries) {
            running += weight
            if (target < running) {
                return letter
            }
        }

        return weightedEntries.last().first
    }

    fun contains(letter: Char): Boolean {
        val uppercase = letter.uppercaseChar()
        return weightedEntries.any { it.first == uppercase }
    }
}

object DefaultLetterWeights {
    val vowels: Map<Char, Int> = mapOf(
        'A' to 15,
        'E' to 21,
        'I' to 13,
        'O' to 13,
        'U' to 5
    )

    val consonants: Map<Char, Int> = mapOf(
        'B' to 2,
        'C' to 3,
        'D' to 6,
        'F' to 2,
        'G' to 3,
        'H' to 2,
        'J' to 1,
        'K' to 1,
        'L' to 5,
        'M' to 4,
        'N' to 8,
        'P' to 4,
        'Q' to 1,
        'R' to 9,
        'S' to 9,
        'T' to 9,
        'V' to 1,
        'W' to 1,
        'X' to 1,
        'Y' to 1,
        'Z' to 1
    )
}

class LetterGenerator(
    vowelWeights: Map<Char, Int> = DefaultLetterWeights.vowels,
    consonantWeights: Map<Char, Int> = DefaultLetterWeights.consonants
) {
    private val vowelPool = WeightedLetterPool(vowelWeights)
    private val consonantPool = WeightedLetterPool(consonantWeights)

    fun generateLetter(kind: LetterKind, random: Random): Char {
        return when (kind) {
            LetterKind.VOWEL -> vowelPool.draw(random)
            LetterKind.CONSONANT -> consonantPool.draw(random)
        }
    }

    fun isValidLetter(letter: Char, kind: LetterKind): Boolean {
        return when (kind) {
            LetterKind.VOWEL -> vowelPool.contains(letter)
            LetterKind.CONSONANT -> consonantPool.contains(letter)
        }
    }
}